[vcpkg scripts] Sourcelink

andy-upnorth opened the pull request:
Adds automatic SourceLink support when building most ports on Windows by extending `vcpkg/scripts` directly:
- Embedded in the initial builds for each port, including their transitive dependencies (C and C++).
- Inline headers are also supported by producing a dynamic sourcelink file, adjusted for the target build folder
- Works with github, gitlab, and bitbucket automatically
  - Optional parameter to `vcpkg_from_git` can be used for other repositories (`RAW_ACCESS_URL`)
- Supports both static and dynamic builds
- Supports both cmake and msbuild targets

This was based upon these previous efforts (first commit on this branch begins with the entirety of 13090):
- https://github.com/microsoft/vcpkg/pull/13090
- https://github.com/microsoft/vcpkg/pull/4884

The original "13090" code was updated to more closely match the current vcpkg coding standards, and then additional features were added to finish out some gaps.

Highlights:
- vcpkg_write_sourcelink_file
  - Called by most of the `vcpkg_from_xxx` helpers
  - Generates sourcelink JSON file for the specific port
  - Attempts to determine the location of installed headers based upon the extracted source (this can be optionally overridden)
- roll-up sourcelink file
  - Two implemenatations: cmake and msbuild
    - cmake is in `vcpkg_add_sourcelink_link_options`, which is automatically called by `add_executable` and `add_library`
    - msbuild is handled via `vcpkg.targets` and invokes a new powershell script to accomplish the same goal
  - These collect and adjust any installed `sourcelink/*.json` files 
    - This critical replacement of `__VCPKG_INSTALLED_TRIPLET_DIR__` with the installed header location is what enables inline content to be correctly referenced
    - A side-benefit of this is that the number of linked sourcelink files from vcpkg is always 1 or 2 (rather than N)
    - the powershell script could also be re-used in other scenarios if needed (such as part of  using the output from `vcpkg export`)

Regarding `vcpkg-cmake` :
- Update: This PR *omits a change* to the `vcpkg-cmake` port in order to simplify the review and delivery of the changes.
  - The changes are now limited to only the `scripts/` subdirectory.
  - Logic is included in `vcpkg_add_sourcelink_link_options` to discover the current target's sourcelink file.
- Originally, it included a small change to the `vcpkg-cmake` port
  -  That set the flag `VCPKG_SOURCELINK_FILE` to the just-generated sourcelink file for the current port, instead of requiring the discovery step.
  - However after discussion we determined that the smaller PR scope would be simpler to integrate for now.

Other notes:
- Regarding Linux:
  - the sourcelink JSON files are also generated on Linux with suitable paths
  - However these are not embedded into any outputs or processed further.  That's an entirely different feature, though entirely plausible
- Regarding other build systems
  - This was developed and tested with CMake and Visual Studio
  - Other build systems might be possible by leveraging the powershell script to roll-up the sourcelink files
- Sourcelink for the Target project
  - This is beyond the scope of `vcpkg` itself because developers of a project would expect to generate a simple sourcelink JSON and separately add it to the link-line of their target applications or libraries.
- Visual Studio - "Require source files to exactly match"
  - You probably will need to un-check this option in Visual Studio at some point, because any patch applied to a header during the vcpkg build will cause this sort of mismatch
  - For example: https://github.com/microsoft/vcpkg/blob/master/ports/fmt/portfile.cmake#L26-L28
- Binary caching
  - This works very well with binary caching, because the original sourcelink JSON file is part of the cached binary
  - Upon extraction, the sourcelink files will be adjusted by the mechanism described above
- Asset cache and/or generated headers
  - Not directly supported by these changes, as this links to the original URLs for the port.
  - A separate mechanism could be imagined where a small proxy could be developed that's backed by vcpkg
  - However that would require an override into `vcpkg_write_sourcelink_file` to establish the long-lived internal URL of this hypothetical proxy service.  A plausible extension, but an additional effort.

andy-upnorth committed:
Fixed mistake in vcpkg.cmake where a flag was not correctly cleared

andy-upnorth committed:
Initial manual merge of changes from previous contributors, as referenced in branch ras0219/vcpkg/tree/dev/lukaszme/sourcelink - part of vcpkg issue 13090

Note: This represents the original code, but manually-merged into the new target location (after 4 years of change).  Explicitly avoided any extra code changes (naming convention, etc) in this initial commit.

andy-upnorth committed:
Updated original changes to align with current vcpkg code style (starting with vcpkg ticket 13090)

andy-upnorth committed:
Removed incorrect escaping for ninja build, which was preventing the sourcelink calls from being added.
Also confirmed that embedded spaces are automatically quoted in the ninja output, so explicit quoting in the cmake code is unnecessary.

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth committed:
in add_library also add sourcelink in IMPORTED case

andy-upnorth committed:
Also allow C projects to use sourcelink

andy-upnorth committed:
Updated Z_CURRENT_SOURCELINK_FILE when building ports to match the path used in vcpkg_write_sourcelink_file, so it can be located.
Note that this information is passed along into the port build, which is subsequently picked up in the invoked build.. such as via buildsystems/vcpkg.cmake

andy-upnorth committed:
Automatically add the current port's newly-created sourcelink file via VCPKG_SOURCELINK_FILE into the build, which will automatically pick up its dependency sourcelink files.

andy-upnorth committed:
Add sourcelink comments to clarify how some of this flows through the tooling

andy-upnorth committed:
Enhanced vcpkg_write_sourcelink_file to allow optional mapping from installed headers to the original repo location, which allows inlined code to be mapped back via sourceLink.
Updated vcpkg_from_github to include default RAW_INCLUDE_MAPPING values and to allow for optional override by callers of ports that opt to provide this extra detail

andy-upnorth committed:
Added option for vcpkg_from_github to specify a different raw url, in case the public github is not being used

andy-upnorth committed:
Cleaned up RAW_INCLUDE_MAPPING to be platform-neutral

andy-upnorth committed:
Add sourcelink generation for gitlab and hosted bitbucket

andy-upnorth committed:
Add sourcelink to vcpkg_from_git, which is triggered by setting RAW_ACCESS_URL appropriate for the git server that is being used.
Examples are provided in the comments within vcpkg_from_git

andy-upnorth committed:
Centralized automatic sourcelink path discovery inside vcpkg_write_sourcelink_file

andy-upnorth committed:
Overhaul vcpkg_add_sourcelink_link_options to parse each of the generated sourcelink JSON files and update the contents with the installed triplet directory.
All of the sourcelink information is combined together into a single JSON file for all ports and their dependencies, including both the original buildtree folder and also the installed header locations.

andy-upnorth committed:
Simplified vcpkg_add_sourcelink_link_options now that it is always combining all the discovered files together into a single JSON file

andy-upnorth committed:
Fixed basic sourcelink usage with msbuild targets (only the buildtree, not adjusted for the installed location)

andy-upnorth committed:
Expanded sourcelink support for msbuild to fully support installed headers, with a new powershell script that combines each of the template sourcelink files and generates a single sourcelink file to auto-link (same as cmake).
The same powershell script could potentially be used in other scenarios where a directory of template vcpkg sourcelink files need to be merged and adjusted for their install location.

andy-upnorth committed:
Updated msbuild powershell invocation to also check for pwsh.exe, like the other example in the same file.
However set the StandardErrorImportance to low in order to hide the complaints about not being found... there's an automatic fallback to the system powershell

andy-upnorth committed:
Suppress the message about missing pwsh.exe by setting StandardErrorImportance to low

andy-upnorth committed:
minor tidying up the msbuild sourcelink comments

andy-upnorth committed:
Cleaned up vcpkg_add_sourcelink_link_options and updated comments

andy-upnorth committed:
fixed typos in error messages

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth committed:
Refined vcpkg_add_sourcelink_link_options to perform an extra seach for sourcelink files if VCPKG_SOURCELINK_FILE was not provided during a port build.

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth commented:
@microsoft-github-policy-service agree

andy-upnorth commented:
@microsoft-github-policy-service agree

andy-upnorth commented on scripts/buildsystems/msbuild/vcpkg.targets:
```diff
@@ -228,6 +286,7 @@
       Condition="'$(VcpkgXUseBuiltInApplocalDeps)' != 'true'"
       Command="pwsh.exe $(_ZVcpkgAppLocalPowerShellCommonArguments)"
       IgnoreExitCode="true"
+      StandardErrorImportance="low"
```
Changing this to `low` actually fixes this issue: https://github.com/microsoft/vcpkg/issues/38502

andy-upnorth commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -656,6 +780,14 @@ function(add_library)
+        if(NOT VCPKG_DISABLE_SOURCELINK)
+            vcpkg_add_sourcelink_link_options("${target_name}")
+        endif()
+    elseif(ALIAS_IDX EQUAL "-1")
```
This was missing in the orginal 13090 version, which is why shared-library ports did not embed any sourcelink information.

andy-upnorth commented on scripts/cmake/vcpkg_configure_cmake.cmake:
```diff
@@ -216,6 +216,10 @@ function(vcpkg_configure_cmake)
     endif()
 
+    if(EXISTS ${Z_CURRENT_SOURCELINK_FILE})
+        vcpkg_list(APPEND arg_OPTIONS "-DVCPKG_SOURCELINK_FILE=${Z_CURRENT_SOURCELINK_FILE}")
```
This file was in the original 13090 pull-request, but since that time this file is actually deprecated -- a quick search of the tip of `ports/` no longer shows any calls to `vcpkg_configure_cmake`.  Feedback welcome.

andy-upnorth commented on scripts/cmake/vcpkg_from_bitbucket.cmake:
```diff
@@ -89,5 +89,24 @@ ${version_contents}
+
+    vcpkg_write_sourcelink_file(
+        SOURCE_PATH "${SOURCE_PATH}"
+        SERVER_PATH "https://bitbucket.com/${org_name}/${repo_name}/raw/${ref_to_use}/*"
```
This is hard-coded to `bitbucket.com`, matching the pattern throughout the rest of the file

andy-upnorth commented on scripts/cmake/vcpkg_write_sourcelink_file.cmake:
```diff
@@ -0,0 +1,152 @@
+  - This is critical for debugging into the library itself, and only depends on the SERVER_PATH being valid
+  - If a Shared library is produced, then this is embedded into the PDB (on Windows)
+  - If a Static library is produced, then the JSON will be embedded in the eventual executable or shared library that uses it
+- The remaining entries are prefixed with `__VCPKG_INSTALLED_TRIPLET_DIR__`, which serves as a placeholder
```
The use of `__VCPKG_INSTALLED_TRIPLET_DIR__` is the crux of enabling inline headers to work correctly.

andy-upnorth commented on scripts/cmake/vcpkg_write_sourcelink_file.cmake:
```diff
@@ -0,0 +1,152 @@
+            set(check_repo_name True)
+        endif()
+
+        # The order is relevant here, so the most deply-nested items or most interesting are found first
```
This is the other critical aspect of inline headers -- some sensible defaults about where to find "most headers" in repositories.  Complex C++ headers are by far the most interesting here, and they often match the patterns below.
Of course this can be overridden, but these work a fair bit of the time.

andy-upnorth commented on ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:
This is the only change to the `vcpkg-cmake` port itself, and there's an alternate approach (a less elegant) inside `vcpkg_add_sourcelink_link_options` that handles the case where this flag is not passed correctly.

I'm looking for feedback from the reviewers on this topic -- is it better to remove this file from the PR so it can be cleanly only the `scripts` directories?   In that case, this change would be potentially merged in a future PR if wanted.

andy-upnorth commented on scripts/cmake/vcpkg_write_sourcelink_file.cmake:
```diff
@@ -0,0 +1,152 @@
+    string(REGEX REPLACE "\\\\" "\\\\\\\\" sourcelink_source_path "${sourcelink_source_path}")
+
+    # Write the first line of the file, which is used for the immediate build of this port
+    file(WRITE "${CURRENT_PACKAGES_DIR}/sourcelink/${PORT}.json" "{\"documents\":{ \"${sourcelink_source_path}\": \"${arg_SERVER_PATH}\"")
```
Question for reviewers:  These files are all written to a *new folder* called `sourcelink` next to `bin`, `share`, etc.
I simply kept the original approach from the previous work on this topic.

However if there is objection to this location, it should be possible to change this to `share/sourcelink/*` or a similar pattern if needed.  The path needs to persist and be used in several subsequent steps, so best to get it the preferred pattern now.

andy-upnorth committed:
Reduced scope of changes to only impact the scripts/ folder by removing the change in the vcpkg-cmake port that provides VCPKG_SOURCELINK_FILE explicitly.
There is no change in functionality, but is a simpler changeset that should more easily pass the automated build tests.

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth commented:
Hi @JavierMatosD, thanks for the review of these changes, and the team's willingness to generally move forward with the feature!

Here are the 2 topics that come to mind:

### Topic 1:  `include/` mappings mentioned in your comment
I re-checked the logic in `vcpkg_write_sourcelink_file` and I believe the default cases handle the situation because it always includes the PORT name as part of the include path. 
There is certainly an opportunity to improve the error-checking of the inputs (prevent `../` and `/` for example).

For example, here are the shortest two snippets that are checked
```
...
        elseif(IS_DIRECTORY  "${arg_SOURCE_PATH}/include")
            list(APPEND raw_include_mapping "${PORT}/*" "include/*")
...
        else()
            # If all else fails, at least map to the top of the repo
            list(APPEND raw_include_mapping "${PORT}/*" "*")
        endif()
```
In the snippets above, you'll notice:
- the first string (local path, embedded in PDB) always has `${PORT}/*` , which is mapped to `..../include/${PORT}/*` 
- The second string with a wildcard string of `include/*` is for the git repo on the server.

Therefore I believe the default cases are handled correctly, but if you see a logic error I've missed then please point it out.


The special case of files in the top-level folder is described in my docs, but I agree does leave more up to the whims of each project:

> This also allows individual files to be overridden, even the uncommon situation of files directly inside the include directory.
> ```cmake
>     list(APPEND raw_include_mapping "${PORT}/myfile.h" "special/myfile.h")
>     list(APPEND raw_include_mapping "${PORT}.h" "${PORT}.h")
> ```

I could envision scenarios where a poorly-formed `RAW_INCLUDE_MAPPING` could lead to issues, so blocking the case of individual files within the include folder would be a viable solution.  That has a minor inconvience of blocking files such as `include/zlib.h` but those rarely have much inlined code anyway. 

_Summary_ : Based on your feedback, I will presume that the top-level include files was the primary concern.  Therefore I will modify the script to block this category of files, requiring that the `RAW_INCLUDE_MAPPING` must always reference a subdirectory.  
Note that I can't directly require `${PORT}` because of special-cases like `boost`, where all the headers are beneath that subtree of `inclulde/boost/xyz` but don't precisely match the `PORT` name itself.

### Topic 2: Location of generated sourcelink/ files
This design always outputs the files to a new folder tree called `sourcelink`, which is at the same level as `bin`, `include`, `share`, and others.  

My question is whether adding a new top-level published folder is an acceptable pattern?
... or if I should try an instead publish these under `share/sourcelink/*.json` instead.

What's the preferred pattern from the vcpkg team?

andy-upnorth commented:
By the way, here is a sample produced on my local system when I used the following vcpkg dependency list for a cmake build:
```
  "dependencies": [
    "fmt",
    "libpng",
    "rapidjson",
    "qtmultimedia",
    "expat"
  ]
```

The merged `vcpkg.sourcelink.json` includes most of the projects above (except `qtmultimedia` itself, because the `RAW_URL` is non-standard and not overridden for that project).  This merged content is built from the 18 smaller 2-entry files produced by each of the port builds.

The merged file (which would be auto-linked into the final project) contains only unique keys because each port typically has only its 2 entries (buildtrees and installed location).
```json
{"documents":{ 
 "D:\\mydev\\vcpkg\\buildtrees\\brotli\\src\\v1.1.0-958bece74c.clean\\*" : "https://raw.githubusercontent.com/google/brotli/v1.1.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\brotli\\*" : "https://raw.githubusercontent.com/google/brotli/v1.1.0/*"
,"D:\\mydev\\vcpkg\\buildtrees\\dbus\\src\\bus-1.16.2-c5722a591d.clean\\*" : "https://gitlab.freedesktop.org//dbus/dbus/-/raw/dbus-1.16.2/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\dbus\\*" : "https://gitlab.freedesktop.org//dbus/dbus/-/raw/dbus-1.16.2/dbus/*"
,"D:\\mydev\\vcpkg\\buildtrees\\double-conversion\\src\\v3.3.0-aed627b50a.clean\\*" : "https://raw.githubusercontent.com/google/double-conversion/v3.3.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\double-conversion\\*" : "https://raw.githubusercontent.com/google/double-conversion/v3.3.0/double-conversion/*"
,"D:\\mydev\\vcpkg\\buildtrees\\egl-registry\\src\\274f9019e6-49d6bdb34a.clean\\*" : "https://raw.githubusercontent.com/KhronosGroup/EGL-Registry/7db3005d4c2cb439f129a0adc931f3274f9019e6/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\egl-registry\\*" : "https://raw.githubusercontent.com/KhronosGroup/EGL-Registry/7db3005d4c2cb439f129a0adc931f3274f9019e6/*"
,"D:\\mydev\\vcpkg\\buildtrees\\expat\\src\\R_2_7_1-4bba8bb538.clean\\*" : "https://raw.githubusercontent.com/libexpat/libexpat/R_2_7_1/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\expat\\*" : "https://raw.githubusercontent.com/libexpat/libexpat/R_2_7_1/*"
,"D:\\mydev\\vcpkg\\buildtrees\\fmt\\src\\11.0.2-c30c0a133f.clean\\*" : "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\fmt\\*" : "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/include/fmt/*"
,"D:\\mydev\\vcpkg\\buildtrees\\freetype\\src\\VER-2-13-3-8b423a038b.clean\\*" : "https://gitlab.freedesktop.org//freetype/freetype/-/raw/VER-2-13-3/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\freetype\\*" : "https://gitlab.freedesktop.org//freetype/freetype/-/raw/VER-2-13-3/include/freetype/*"
,"D:\\mydev\\vcpkg\\buildtrees\\harfbuzz\\src\\10.2.0-95c1fd1d78.clean\\*" : "https://raw.githubusercontent.com/harfbuzz/harfbuzz/10.2.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\harfbuzz\\*" : "https://raw.githubusercontent.com/harfbuzz/harfbuzz/10.2.0/src/*"
,"D:\\mydev\\vcpkg\\buildtrees\\libjpeg-turbo\\src\\3.1.0-61083179ee.clean\\*" : "https://raw.githubusercontent.com/libjpeg-turbo/libjpeg-turbo/3.1.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\libjpeg-turbo\\*" : "https://raw.githubusercontent.com/libjpeg-turbo/libjpeg-turbo/3.1.0/src/*"
,"D:\\mydev\\vcpkg\\buildtrees\\libpng\\src\\v1.6.46-0d105c43f9.clean\\*" : "https://raw.githubusercontent.com/pnggroup/libpng/v1.6.46/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\libpng\\*" : "https://raw.githubusercontent.com/pnggroup/libpng/v1.6.46/*"
,"D:\\mydev\\vcpkg\\buildtrees\\lz4\\src\\v1.10.0-61bd08d80e.clean\\*" : "https://raw.githubusercontent.com/lz4/lz4/v1.10.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\lz4\\*" : "https://raw.githubusercontent.com/lz4/lz4/v1.10.0/*"
,"D:\\mydev\\vcpkg\\buildtrees\\opengl-registry\\src\\f632f7ea33-4952f9ace9.clean\\*" : "https://raw.githubusercontent.com/KhronosGroup/OpenGL-Registry/3530768138c5ba3dfbb2c43c830493f632f7ea33/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\opengl-registry\\*" : "https://raw.githubusercontent.com/KhronosGroup/OpenGL-Registry/3530768138c5ba3dfbb2c43c830493f632f7ea33/*"
,"D:\\mydev\\vcpkg\\buildtrees\\openssl\\src\\nssl-3.5.0-b8228451bd.clean\\*" : "https://raw.githubusercontent.com/openssl/openssl/openssl-3.5.0/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\openssl\\*" : "https://raw.githubusercontent.com/openssl/openssl/openssl-3.5.0/include/openssl/*"
,"D:\\mydev\\vcpkg\\buildtrees\\pcre2\\src\\e44fa74ca4-bc7898d1f1.clean\\*" : "https://raw.githubusercontent.com/zherczeg/sljit/8481dde366d0346ac5475aa03ae48ee44fa74ca4/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\pcre2\\*" : "https://raw.githubusercontent.com/zherczeg/sljit/8481dde366d0346ac5475aa03ae48ee44fa74ca4/*"
,"D:\\mydev\\vcpkg\\buildtrees\\pkgconf\\src\\conf-2.4.3-0e28407cb4.clean\\*" : "https://raw.githubusercontent.com/pkgconf/pkgconf/pkgconf-2.4.3/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\pkgconf\\*" : "https://raw.githubusercontent.com/pkgconf/pkgconf/pkgconf-2.4.3/*"
,"D:\\mydev\\vcpkg\\buildtrees\\rapidjson\\src\\106cf7102f-b0877c68bd.clean\\*" : "https://raw.githubusercontent.com/Tencent/rapidjson/24b5e7a8b27f42fa16b96fc70aade9106cf7102f/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\rapidjson\\*" : "https://raw.githubusercontent.com/Tencent/rapidjson/24b5e7a8b27f42fa16b96fc70aade9106cf7102f/include/rapidjson/*"
,"D:\\mydev\\vcpkg\\buildtrees\\zlib\\src\\v1.3.1-2e5db616bf.clean\\*" : "https://raw.githubusercontent.com/madler/zlib/v1.3.1/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\zlib\\*" : "https://raw.githubusercontent.com/madler/zlib/v1.3.1/*"
,"D:\\mydev\\vcpkg\\buildtrees\\zstd\\src\\v1.5.7-bb6cae2b2e.clean\\*" : "https://raw.githubusercontent.com/facebook/zstd/v1.5.7/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\zstd\\*" : "https://raw.githubusercontent.com/facebook/zstd/v1.5.7/*"
}}
```

JavierMatosD commented:
Hi Andy — thanks again for the detailed clarifications! 

Topic 1 (include/ mapping):

Yes — I believe the direction you described is correct. We want to avoid fallback mappings to `include/*`. If no unambiguous subdirectory (such as `include/${repo_name}`) is found, then:

✅  If the port author explicitly provides RAW_INCLUDE_MAPPING, that can be used — it is the author’s responsibility to ensure mappings are correct and non-conflicting.

❌ If no RAW_INCLUDE_MAPPING is provided, the script should skip generating a mapping — no fallback to include/*.

Topic 2 (folder location):

Per the official vcpkg installation tree layout [here](https://learn.microsoft.com/en-us/vcpkg/reference/installation-tree-layout), the `share/` folder is intended for per-port metadata that is not used at runtime. Since the SourceLink JSON files fit this category, we recommend placing them under:

`share/sourcelink/`

This keeps the layout consistent with vcpkg patterns and avoids introducing new root-level folders.

andy-upnorth committed:
Address code review feedback regarding automatically-generated the RAW_INCLUDE_MAPPING and clarified comments regarding the behavior

andy-upnorth committed:
Sourcelink files are generated into share/sourcelink/PORT to align with the standard vcpkg pattern.

andy-upnorth committed:
Merge branch 'master' into sourcelink

andy-upnorth commented:
I have addressed the feedback from @JavierMatosD and sync'd from vcpkg master.

One upshot of removing the default fallback to `include/PORT/*` is that the combined sourcelink is smaller - containing only the entries that are very likely to work.   I will attach the requested screenshots for manual verification soon.

Output from a trivial cmake-based project:
```
D:\mydev\srclinkExperiment>c:\SrcSrvSDK\pdbstr.exe -r -p:"D:\mydev\srclinkExperiment\build\HelloWorld.pdb" -s:sourcelink$1
{"documents":{
 "D:\\mydev\\vcpkg\\buildtrees\\expat\\src\\R_2_7_1-4bba8bb538.clean\\*" : "https://raw.githubusercontent.com/libexpat/libexpat/R_2_7_1/*"
,"D:\\mydev\\vcpkg\\buildtrees\\fmt\\src\\11.0.2-c30c0a133f.clean\\*" : "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\fmt\\*" : "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/include/fmt/*"
,"D:\\mydev\\vcpkg\\buildtrees\\libpng\\src\\v1.6.46-0d105c43f9.clean\\*" : "https://raw.githubusercontent.com/pnggroup/libpng/v1.6.46/*"
,"D:\\mydev\\vcpkg\\buildtrees\\rapidjson\\src\\106cf7102f-b0877c68bd.clean\\*" : "https://raw.githubusercontent.com/Tencent/rapidjson/24b5e7a8b27f42fa16b96fc70aade9106cf7102f/*"
,"D:\\mydev\\srclinkExperiment\\build\\vcpkg_installed\\x64-windows\\include\\rapidjson\\*" : "https://raw.githubusercontent.com/Tencent/rapidjson/24b5e7a8b27f42fa16b96fc70aade9106cf7102f/include/rapidjson/*"
,"D:\\mydev\\vcpkg\\buildtrees\\zlib\\src\\v1.3.1-2e5db616bf.clean\\*" : "https://raw.githubusercontent.com/madler/zlib/v1.3.1/*"
}}
```

Also confirmed that it continues to work from a vcxproj solution build (leveraging the powershell script) from the new `share/sourcelink` location:
```
{"documents": {
    "D:\\mydev\\vcpkg\\buildtrees\\boost-function\\src\\ost-1.86.0-ec6eb5e976.clean\\*": "https://raw.githubusercontent.com/boostorg/function/boost-1.86.0/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\tuple\\*": "https://raw.githubusercontent.com/boostorg/tuple/boost-1.86.0/include/boost/tuple/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\intrusive\\*": "https://raw.githubusercontent.com/boostorg/intrusive/boost-1.86.0/include/boost/intrusive/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-mp11\\src\\ost-1.86.0-93cf4ea149.clean\\*": "https://raw.githubusercontent.com/boostorg/mp11/boost-1.86.0/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\utility\\*": "https://raw.githubusercontent.com/boostorg/utility/boost-1.86.0/include/boost/utility/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\mp11\\*": "https://raw.githubusercontent.com/boostorg/mp11/boost-1.86.0/include/boost/mp11/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-move\\src\\ost-1.86.0-a83c27c31c.clean\\*": "https://raw.githubusercontent.com/boostorg/move/boost-1.86.0/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\io\\*": "https://raw.githubusercontent.com/boostorg/io/boost-1.86.0/include/boost/io/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\move\\*": "https://raw.githubusercontent.com/boostorg/move/boost-1.86.0/include/boost/move/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\fusion\\*": "https://raw.githubusercontent.com/boostorg/fusion/boost-1.86.0/include/boost/fusion/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-config\\src\\ost-1.86.0-8ca7c2ca3e.clean\\*": "https://raw.githubusercontent.com/boostorg/config/boost-1.86.0/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-function-types\\src\\ost-1.86.0-aa3ea9a2df.clean\\*": "https://raw.githubusercontent.com/boostorg/function_types/boost-1.86.0/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-functional\\src\\ost-1.86.0-1c8b845386.clean\\*": "https://raw.githubusercontent.com/boostorg/functional/boost-1.86.0/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-static-assert\\src\\ost-1.86.0-e7e646c10a.clean\\*": "https://raw.githubusercontent.com/boostorg/static_assert/boost-1.86.0/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-container\\src\\ost-1.86.0-bbec49e1ca.clean\\*": "https://raw.githubusercontent.com/boostorg/container/boost-1.86.0/*",
    "D:\\mydev\\vcpkg\\buildtrees\\rapidjson\\src\\c9dd6067e5-8a215fc499.clean\\*": "https://raw.githubusercontent.com/Tencent/rapidjson/a95e013b97ca6523f32da23f5095fcc9dd6067e5/*",
    "D:\\mydev\\vcpkg\\buildtrees\\zlib\\src\\v1.3.1-2e5db616bf.clean\\*": "https://raw.githubusercontent.com/madler/zlib/v1.3.1/*",
    "D:\\mydev\\vcpkg\\buildtrees\\fmt\\src\\11.0.2-c30c0a133f.clean\\*": "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\fmt\\*": "https://raw.githubusercontent.com/fmtlib/fmt/11.0.2/include/fmt/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\integer\\*": "https://raw.githubusercontent.com/boostorg/integer/boost-1.86.0/include/boost/integer/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\boost\\bind\\*": "https://raw.githubusercontent.com/boostorg/bind/boost-1.86.0/include/boost/bind/*",
    "D:\\mydev\\srclinkExperimentSln\\HelloSrcLink\\vcpkg_installed\\x64-windows\\x64-windows\\include\\rapidjson\\*": "https://raw.githubusercontent.com/Tencent/rapidjson/a95e013b97ca6523f32da23f5095fcc9dd6067e5/include/rapidjson/*",
    "D:\\mydev\\vcpkg\\buildtrees\\boost-mpl\\src\\ost-1.86.0-92630279e5.clean\\*": "https://raw.githubusercontent.com/boostorg/mpl/boost-1.86.0/*",

...

```

andy-upnorth commented:
As confirmation that these changes work as expected, @JavierMatosD  requested the following:

> Lastly, since this behavior can’t be validated automatically, we’d like to see screenshots attached to the PR demonstrating the following scenario as a manual acceptance test:
> 
> Install zlib (unmodified port)
> Purge buildtrees/
> Build a test app using zlib
> Launch under the debugger (F5)
> Step into zlibVersion() (or another non-generated symbol)
> Visual Studio should fetch and show the correct source file (e.g. zlib.c)
> Note: the specific port used doesn’t matter—we just want to confirm the integration works end-to-end.

I followed the standard vcpkg HelloWorld project, then added the following dependencies:
```
  "dependencies": [
    "fmt",
    "libpng",
    "rapidjson",
    "expat"
  ]
```
In order to illustrate stepping into code:
- parsed a `rapidjson::Document` using its `parse` method (header-only),
- a custom formatter calls into `fmt::format_to` ,
- called into `png_get_copyright`

The entire code for the sample is here:
```
#include <fmt/core.h>
#include <png.h>
#include <rapidjson/document.h>

struct HelloStruct
{
    int hello = 42;
    std::string world = "around and around";
};

template <> struct fmt::formatter<HelloStruct, char>
{
    constexpr auto parse(format_parse_context& ctx)
    {
        return ctx.begin();
    }

    template <typename FormatContext> auto format(const HelloStruct& val, FormatContext& ctx) const
    {
        return fmt::format_to(ctx.out(), "{}_{}", val.hello, val.world);
    }
};

int main()
{
    fmt::print("Hello World!\n");

    std::string json = R"(
        {
            "hello": "world"
        }
    )";

    rapidjson::Document doc;
    doc.Parse(json.c_str());
    fmt::print("hello: {}\n", doc["hello"].GetString());

    HelloStruct helloStruct;
    fmt::print("message: {}\n", helloStruct);

    fmt::print("libpng: {}\n", png_get_copyright(nullptr));
   
    return 0;
}
```

After building this, I copied the PDBs into the same output folder as my executable, then removed the vcpkg `buildtrees` and other build-time items.  The attached screenshots show the downloaded files from the Visual Studio cache rather than vcpkg. 

Step into RapidJSON parse (header-only library)
![image](https://github.com/user-attachments/assets/06b8df3a-82b8-4891-b347-11ed526e1bfe)

Breakpoint in format callback - stepping to inlined code:
![image](https://github.com/user-attachments/assets/cd359870-ce92-47f5-be75-80f1759e48ef)

Breakpoint in format callback - jumping to code from DLL (fmtd.dll):
![image](https://github.com/user-attachments/assets/827c638a-920b-4b9d-8501-1df2ed47f633)

Stepping into libpng
![image](https://github.com/user-attachments/assets/d743291b-a2cd-46ee-a9a8-df26ffb952ab)

I have also confirmed that this works on another dev system that doesn't have vcpkg installed at all (only VS2022).

andy-upnorth commented:
I see that one of items in `x64_linux` failed, which turns out is due to a possible mistake in its port, where it seems to mis-use the directory `share` for something else.
https://github.com/microsoft/vcpkg/blob/master/ports/azure-kinect-sensor-sdk/portfile.cmake#L48

Is that typically handled by creating another small PR to fix that item, merging it, then syncing to this?

andy-upnorth commented:
Waiting for https://github.com/microsoft/vcpkg/pull/46022 to be merged,  then will try this again

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

JavierMatosD commented:
Hi Andy,

Thanks for the detailed update and for attaching those screenshots — they’re super helpful. I’ve reviewed the test output and confirmed that the debugger behavior looks correct in all cases. This is exactly what we were hoping to see.

The current CI failures (e.g., with qtmultimedia and thrift) do not appear related to your changes. I’ll investigate those further on our side — no action is needed from you there for now.

My next step is to complete the set of smoke test scenarios we outlined internally (e.g., static vs. shared builds, header-only libraries, conflicting includes, etc.) so we can summarize what this SourceLink setup supports and where the boundaries are. I’ll aim to have that wrapped up shortly.

Appreciate you sticking with this through the final stretch!

Best,
Javier

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth commented:
I have re-sync'd the branch, and there are a few new failures... because it's been so long waiting to get this merged.
Happily, the two older failures from June 2025 seem to have been fixed.


The following are failing on x64_windows:
```
REGRESSION: log4cxx:x64-windows failed with BUILD_FAILED. If expected, add log4cxx:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
REGRESSION: popsift:x64-windows failed with BUILD_FAILED. If expected, add popsift:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
REGRESSION: subversion:x64-windows failed with BUILD_FAILED. If expected, add subversion:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
REGRESSION: vsgxchange:x64-windows failed with BUILD_FAILED. If expected, add vsgxchange:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
```

On Linux and Android, only this one fails:
```
REGRESSION: vsgxchange:x64-linux failed with BUILD_FAILED. If expected, add vsgxchange:x64-linux=fail to /vcpkg/scripts/azure-pipelines/../ci.baseline.txt.
REGRESSION: vsgxchange:x64-android failed with BUILD_FAILED. If expected, add vsgxchange:x64-android=fail to /vcpkg/scripts/azure-pipelines/../ci.baseline.txt.
```

On windows, the error about vsgxchange is:
```
Building vsgxchange:x64-windows@1.1.9...
...
-- Configuring x64-windows
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/bin/cmake.exe --build . --config Debug --target install -- -v -j33
    Working Directory: D:/b/vsgxchange/x64-windows-dbg
    See logs for more information:
      D:\b\vsgxchange\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/vsgxchange/portfile.cmake:32 (vcpkg_cmake_install)
  scripts/ports.cmake:207 (include)
```

Similarly, log4cxx:
```
Installing 1645/2683 log4cxx:x64-windows@1.6.1...
...
-- Configuring x64-windows
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/bin/cmake.exe --build . --config Debug --target install -- -v -j33
    Working Directory: D:/b/log4cxx/x64-windows-dbg
    See logs for more information:
      D:\b\log4cxx\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/log4cxx/portfile.cmake:25 (vcpkg_cmake_install)
  scripts/ports.cmake:207 (include)
```

Also subversion:
```
Installing 2364/2700 subversion:x86-windows@2025-12-23...
...
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: D:\\downloads\\tools\\ninja-1.13.2-windows\\ninja.exe -v
    Working Directory: D:/b/subversion/x86-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      D:\b\subversion\config-x86-windows-dbg-CMakeCache.txt.log
      D:\b\subversion\config-x86-windows-rel-CMakeCache.txt.log
      D:\b\subversion\config-x86-windows-dbg-CMakeConfigureLog.yaml.log
      D:\b\subversion\config-x86-windows-rel-CMakeConfigureLog.yaml.log
      D:\b\subversion\config-x86-windows-out.log

Call Stack (most recent call first):
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:269 (vcpkg_execute_required_process)
  ports/subversion/portfile.cmake:16 (vcpkg_cmake_configure)
  scripts/ports.cmake:207 (include)
```

And popsift
```
Building popsift:x64-windows@0.10.0...
...
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/bin/cmake.exe --build . --config Debug --target install -- -v -j33
    Working Directory: D:/b/popsift/x64-windows-dbg
    See logs for more information:
      D:\b\popsift\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/popsift/portfile.cmake:30 (vcpkg_cmake_install)
  scripts/ports.cmake:207 (include)
```

petersteneteg commented:
You can download the logs here:
https://dev.azure.com/vcpkg/public/_build/results?buildId=125915&view=artifacts&pathAsName=false&type=publishedArtifacts
if you hover a line there is a "..." button on the right with a download option

subversion:
```
CMake Error at D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/share/cmake-3.31/Modules/FindPkgConfig.cmake:938 (message):
  None of the required 'apr-2;apr-1' found
Call Stack (most recent call first):
  CMakeLists.txt:266 (pkg_search_module)


CMake Error at D:/a/_work/1/s/scripts/buildsystems/vcpkg.cmake:779 (_add_library):
  _add_library cannot create ALIAS target "external-apr" because target
  "PkgConfig::apr" does not already exist.
Call Stack (most recent call first):
  CMakeLists.txt:267 (add_library)


-- Checking for module 'apr-util-1'
--   Package 'apr-util-1' not found
CMake Error at D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/share/cmake-3.31/Modules/FindPkgConfig.cmake:645 (message):
  The following required packages were not found:

   - apr-util-1

Call Stack (most recent call first):
  D:/downloads/tools/cmake-3.31.10-windows/cmake-3.31.10-windows-x86_64/share/cmake-3.31/Modules/FindPkgConfig.cmake:873 (_pkg_check_modules_internal)
  CMakeLists.txt:271 (pkg_check_modules)
```

log4cxx:
```
asyncbuffer.cpp.obj : error LNK2019: unresolved external symbol "void __cdecl fmt::v12::assert_fail(char const *,int,char const *)" (?assert_fail@v12@fmt@@YAXPEBDH0@Z) referenced in function "public: void __cdecl fmt::v12::detail::bigint::operator=<int>(int)" (??$?4H@bigint@detail@v12@fmt@@QEAAXH@Z)
```

vsgxchange:
```
D:\b\vsgxchange\src\v1.1.9-935fb40163.clean\src\ktx\ktx.cpp(21): fatal error C1083: Cannot open include file: 'ktxvulkan.h': No such file or directory
```

popsift:
```
C:\PROGRA~1\NVIDIA~2\CUDA\v12.9\bin\crt\link.stub(187): fatal error C1068: cannot open file ':D:/b/popsift/x64-windows-dbg/CMakeFiles/vcpkg.sourcelink.json'
```

Only popsift seems to be related to this PR?

andy-upnorth committed:
Merge branch 'microsoft:master' into sourcelink

andy-upnorth commented:
I've confirmed locally that the popsift port can be fixed by defining `VCPKG_DISABLE_SOURCELINK` to this block in that project's [portfile.cmake](https://github.com/microsoft/vcpkg/blob/fe73195d950e87f0bb6bc91da8e9d52f16d5abac/ports/popsift/portfile.cmake#L24-L28):

This failed for `x64_windows_release` and not on the non-windows builds.  It appears that all of the other build failures (so far) are pre-existing build issues not caused by this PR.  All of the non-Windows builds have finished, but I'm waiting for the rest to complete.

```
vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS ${FEATURE_OPTIONS}
        "-DCMAKE_CUDA_COMPILER=${NVCC_PATH}"
        "-DVCPKG_DISABLE_SOURCELINK=TRUE"
)

```

I believe the reason it fails is because the `/sourcelink:` parameter is passed to the CUDA linker, which doesn't support that flag.  This case was generally supposed to be caught by checking for `MSVC` (from the original implementor of this sourcelink change)... however using the flag to supporess sourcelink for this single project does work.

If this is an acceptable approach, then I can open a PR for the popsift vcpkg port with this added flag.
